### ISSUE TITLE
Support custom Client/Dialect

### ISSUE THREAD

**Issue by vellotis** [tgriesser/knex#1424]
> :information_source: Separate issue for the feature requested in #1280.

It would be great if we could use our own defined dialect which is a instance of base [Client (src/client.js)](https://github.com/tgriesser/knex/blob/e7a42abbb200aaad3f6ac0324c28af75931c0583/src/client.js).

By providing custom dialect in initialization config the suggested fix would be... [src/inrex.js#L17-L22](https://github.com/tgriesser/knex/blob/f37824a11efcb398e11425905fb5bbc864ae8b5a/src/index.js#L17-L22)
```js
/*17*/  if (arguments.length === 0 || (!config.client && !config.dialect)) {
/*18*/    Dialect = makeClient(Client)
/*++*/  } else if (typeof config.client === 'function' && config.client.prototype instanceof Client) {
/*++*/    Dialect = makeClient(config.client)
/*19*/  } else {
/*20*/    var clientName = config.client || config.dialect
/*21*/    Dialect = makeClient(require('./dialects/' + (aliases[clientName] || clientName) + '/index.js'))
/*22*/  }
```
Is the solution acceptable? Test(s) + PR?

BR

**Comment by rhys-vdw** [comment]
@vellotis One thing to point out - you can't rely on `client instanceof Client`, as it may be inheriting from a different version of Knex! (eg. a different node module).

eg.

```js
// my-custom-client.js
var Client = require('knex/lib/client'); // knex 0.10
```

```js
// my-project-in-an-entirely-different-module
var Client = require('knex/lib/client') // knex 0.11
var CustomClient = require('my-custom-client');

assert(CustomClient instanceof Client);
// FAIL! `CustomClient` inherits from the Knex dependency under the `my-custom-client` module.
```

So instead either use [existing methods](https://github.com/tgriesser/knex/blob/e7a42abbb200aaad3f6ac0324c28af75931c0583/src/client.js#L263-L265):

```js
// Check already defined `toString`
if (client.toString() === '[object KnexClient]') {
  makeClient(client);
 }
```

The above method will fail if the new client overrides `toString`. A sentinel prop is a bit more robust:

```js
// knex/src/client.js

// Define some safe property by which we'll identify `Client` types.
const CLIENT_SENTINEL = '__KNEX_CLIENT__';

// Assign it to prototype (so is present on all instances).
Client.prototype[CLIENT_SENTINEL] = true;

// Create helper function on `Client`
Client.prototype.isClient = function (maybeClient) {
  return !!(maybeClient && maybeClient[CLIENT_SENTINEL]);
}

// Using it in knex initialization:
if (Client.isClient(config.client)) {
  makeKnex(config.client);
}
```

**Comment by vellotis** [comment]
@rhys-vdw  I see your point.
But shouldn't the plugin-client always use the Client of currently installed knex version (peerDependency)? As the knex client API for eg. v0.15 may change a fracture amount. Then the plugin-client should not be attachable?

It's just a thought... Otherwise your suggested solution seems to be legit.

**Comment by rhys-vdw** [comment]
Yep. I was wrong. Plugins can use `peerDependencies` locking down the version of Knex with which they're compatible. This will fix the problem I was mentioning.

**Comment by vellotis** [comment]
Closed by PR #1428
